### PR TITLE
added firehouse context to the /consensus

### LINKED ISSUE #14: Instrument consensus
# Description

Instrument the consensus/ functions and object to record all the necessary block/transaction changes as per streamingFast practices. Refer to the streamingFast go-ethereum repository for the required changes.

### PR DESCRIPTION
**Pull request by ohmpatel1997**
Closes #14 
This PR adds the firehouse context to the consensus pkg

### CODE REVIEW

File: consensus/aura/aura.go [hunk 1]
```diff
@@ -825,24 +826,25 @@ func (c *AuRa) Initialize(config *chain.Config, chain consensus.ChainHeaderReade
 
 }
 
-func (c *AuRa) ApplyRewards(header *types.Header, state *state.IntraBlockState, syscall consensus.SystemCall) error {
+func (c *AuRa) ApplyRewards(header *types.Header, state *state.IntraBlockState, syscall consensus.SystemCall, firehoseContext *firehose.Context) error {
 	beneficiaries, _, rewards, err := calculateRewards(c, header, syscall)
```
Comment (dhyaniarun1993) [inline consensus/aura/aura.go]: calculate reward returns the reward kind list as second argument. You can use this to find the `addBalance` reason.

Something like this
```
                switch rewardKinds[i] {
		case aurainterfaces.RewardAuthor:
			state.AddBalance(beneficiaries[i], rewards[i], false, firehoseContext, firehose.BalanceChangeReason("reward_mine_block"))
		case aurainterfaces.RewardUncle:
			state.AddBalance(beneficiaries[i], rewards[i], false, firehoseContext, firehose.BalanceChangeReason("reward_mine_uncle"))
		default:
			state.AddBalance(beneficiaries[i], rewards[i], false, firehoseContext, firehose.BalanceChangeReason("reward_unknown"))
		}
```

File: consensus/parlia/parlia.go [hunk 2]
```diff
@@ -1275,6 +1278,7 @@ func (p *Parlia) systemCall(from, contract libcommon.Address, data []byte, ibs *
 		msg.Gas(),
 		msg.Value(),
 		false,
+		firehoseContext,
```
Comment (dhyaniarun1993) [inline consensus/parlia/parlia.go]: Any reason why this is added?

File: consensus/parlia/parlia.go [hunk 3]
```diff
@@ -615,7 +616,7 @@ func (p *Parlia) VerifyUncles(chain consensus.ChainReader, header *types.Header,
 
 // Prepare initializes the consensus fields of a block header according to the
 // rules of a particular engine. The changes are executed inline.
-func (p *Parlia) Prepare(chain consensus.ChainHeaderReader, header *types.Header, ibs *state.IntraBlockState) error {
+func (p *Parlia) Prepare(chain consensus.ChainHeaderReader, header *types.Header, ibs *state.IntraBlockState, firehoseContext *firehose.Context) error {
```
Comment (dhyaniarun1993) [inline consensus/parlia/parlia.go:619]: You need to update the consensus interface `Prepare` definition along with all implementation of it.

**Review by dhyaniarun1993 — CHANGES_REQUESTED** [review]

File: consensus/parlia/parlia.go [hunk 4]
```diff
@@ -1059,15 +1060,17 @@ func (p *Parlia) getCurrentValidators(header *types.Header, ibs *state.IntraBloc
...
 	coinbase := header.Coinbase
 	balance := state.GetBalance(consensus.SystemAddress).Clone()
 	if balance.Cmp(u256.Num0) <= 0 {
 		return txs, systemTxs, receipts, nil
 	}
-	state.SetBalance(consensus.SystemAddress, u256.Num0)
-	state.AddBalance(coinbase, balance)
+
+	// CS TODO: confirm the message
+	state.SetBalance(consensus.SystemAddress, u256.Num0, firehoseContext, firehose.BalanceChangeReason("ignored"))
```
Comment (dhyaniarun1993) [inline consensus/parlia/parlia.go]: Please use the constant `firehose.IgnoredBalanceChangeReason`

File: consensus/parlia/parlia.go (hunk 2 above)
Comment (ohmpatel1997) [reply consensus/parlia/parlia.go]: Aah, that was added coz I also made dependent changes in other pkg, but then realised i was not supposed to make that, so forgot to remove this. 

Removed.

File: consensus/parlia/parlia.go [hunk 5]
```diff
@@ -1059,15 +1060,17 @@ func (p *Parlia) getCurrentValidators(header *types.Header, ibs *state.IntraBloc
...
+	state.AddBalance(coinbase, balance, false, firehoseContext, firehose.BalanceChangeReason("ignored"))
```
Comment (dhyaniarun1993) [inline consensus/parlia/parlia.go]: I think this should be `firehose.BalanceChangeReason("reward_mine_block")`

**Review by sadiq1971 — APPROVED** [review]

File: consensus/parlia/parlia.go (hunk 4 above)
Comment (ohmpatel1997) [reply consensus/parlia/parlia.go]: actually I did coz it will not match the regex then. But thats fine I think after reading the comments.

**Review by dhyaniarun1993 — APPROVED** [review]

**Pull request merged**